src/backend/distributed/commands/multi_copy.c

```diff
@@ -2118,10 +2137,49 @@ ShouldExecuteCopyLocally(bool isIntermediateResult)
 
 		return true;
 	}
+	else if (IsMultiStatementTransaction())
+	{
+		return true;
+	}
+	else if (!ShardIntervalListHasLocalPlacements(shardIntervalList))
+	{
+		return false;
+	}
+
+	/*
+	 * If we can reserve a connection, we should go ahead with
+	 * remote execution.
+	 *
+	 * NB: It is not advantageous to use remote execution just with a
+	 * single remote connection. In other words, a single remote connection
+	 * would not perform better than local execution. However, we prefer to
+	 * do this because it is likely that the COPY would get more connections
+	 * to parallelize the operation. In the future, we might relax this
+	 * requirement and failover to local execution as on connection attempt
+	 * failures as the executor does.
+	 */
+	return !TryConnectionPossibilityForLocalPrimaryNode();
```

Comment:
it seems a bit strange to have a function with side effects in this ShouldXXX function. Maybe we merge LocalCopyAllowed and ShouldExecuteCopyLocally, but then separate out this part?